pkg/controllers/clusterresourceplacement/controller.go

```diff
@@ -455,38 +478,113 @@ func (r *Reconciler) getOrCreateClusterResourceSnapshot(ctx context.Context, crp
 		return nil, err
 	}
 
-	// create a new resource snapshot
-	latestResourceSnapshotIndex++
-	latestResourceSnapshot = &fleetv1beta1.ClusterResourceSnapshot{
+	// only increment if new cluster resource snapshots are to be created.
+	if shouldCreateNewSnapshot {
+		latestResourceSnapshotIndex++
+	}
+	// split selected resources as list of lists.
+	selectedResourcesList := splitSelectedResources(resourceSnapshotSpec.SelectedResources)
+	// create master cluster resource snapshot and sub-indexed resource snapshots belonging to the same group index.
+	resourceSnapshotSubIndex := -1
+	var resourceSnapshot *fleetv1beta1.ClusterResourceSnapshot
+	for i := 0; i < len(selectedResourcesList); i++ {
+		if i == 0 {
+			resourceSnapshot = buildMasterClusterResourceSnapshot(latestResourceSnapshotIndex, len(selectedResourcesList), envelopeObjCount, crp.Name, resourceHash, selectedResourcesList[i])
+			latestResourceSnapshot = resourceSnapshot
+		} else {
+			resourceSnapshotSubIndex++
+			resourceSnapshot = &fleetv1beta1.ClusterResourceSnapshot{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: fmt.Sprintf(fleetv1beta1.ResourceSnapshotNameWithSubindexFmt, crp.Name, latestResourceSnapshotIndex, resourceSnapshotSubIndex),
+					Labels: map[string]string{
+						fleetv1beta1.CRPTrackingLabel:   crp.Name,
+						fleetv1beta1.ResourceIndexLabel: strconv.Itoa(latestResourceSnapshotIndex),
+					},
+					Annotations: map[string]string{
+						fleetv1beta1.SubindexOfResourceSnapshotAnnotation: strconv.Itoa(resourceSnapshotSubIndex),
+					},
+				},
+				Spec: fleetv1beta1.ResourceSnapshotSpec{
+					SelectedResources: selectedResourcesList[i],
+				},
+			}
+		}
+		if err = r.createResourceSnapshot(ctx, crp, resourceSnapshot); err != nil {
+			return nil, err
+		}
+	}
+	if len(selectedResourcesList) == 0 {
+		resourceSnapshot = buildMasterClusterResourceSnapshot(latestResourceSnapshotIndex, 1, envelopeObjCount, crp.Name, resourceHash, []fleetv1beta1.ResourceContent{})
+		latestResourceSnapshot = resourceSnapshot
+		if err = r.createResourceSnapshot(ctx, crp, resourceSnapshot); err != nil {
+			return nil, err
+		}
+	}
+	return latestResourceSnapshot, nil
+}
+
+// splitSelectedResources splits selected resources in a ClusterResourcePlacement into separate lists
+// so that the total size of each split list of selected Resources is within 1MB limit.
+func splitSelectedResources(selectedResources []fleetv1beta1.ResourceContent) [][]fleetv1beta1.ResourceContent {
```

Comment:
we should add UT for this important function.

Reply:
As discussed offline opened a separate PR to add UTs https://github.com/Azure/fleet/pull/652